### PR TITLE
feat(module): add cloudbuild connection module

### PR DESCRIPTION
**Pull request by caetano-colin**
This PR will add the module `cloudbuild_repo_connection`.

This module is designed to create and manage secret versions, as well as configure the necessary permissions for cloud build service agent when utilizing Cloud Build repositories (2nd generation).

Additionally, it will establish the corresponding Cloud Build repositories based on the `cloudbuild_repos` variable, where users can specify the repository names and URLs from their version control systems.

Users will provide the required secrets through the `credential_config` variable, indicating their chosen Git provider. Currently, the module supports both GitHub and GitLab.

### CODE REVIEW

File: examples/cloudbuild_repo_connection_github/variables.tf [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  type        = string
+}
+
+variable "github_app_id" {
+  description = "The application ID for the Cloudbuild GitHub app."
+  type        = string
+}
+
+variable "test_repo_url" {
+  description = "The HTTPS clone URL of the test repository, ending with .git."
```
Comment (daniel-cit) [inline examples/cloudbuild_repo_connection_github/variables.tf:33]:
```suggestion
variable "repository_url" {
  description = "The HTTPS clone URL of the repository, ending with .git."
```

File: examples/cloudbuild_repo_connection_github/variables.tf [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  type        = string
+}
+
+variable "test_repo_name" {
+  description = "The name of the test repository."
```
Comment (daniel-cit) [inline examples/cloudbuild_repo_connection_github/variables.tf:38]:
```suggestion
variable "repository_name" {
  description = "The name of the repository."
```

File: modules/cloudbuild_repo_connection/variables.tf [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        var.credential_config.github_app_id == null &&
+        var.credential_config.gitlab_read_authorizer_credential != null &&
+        var.credential_config.gitlab_authorizer_credential != null
+      ) : false
+    )
+    error_message = "You must specify a valid credential_type ('GITHUBv2' or 'GITLABv2'). For 'GITHUBv2', all 'github_' prefixed variables must be defined and no 'gitlab_' prefixed variables should be defined. For 'GITLABv2', all 'gitlab_' prefixed variables must be defined and no 'github_' prefixed variables should be defined."
+  }
+}
+
+variable "cloudbuild_repos" {
```
Comment (daniel-cit) [inline modules/cloudbuild_repo_connection/variables.tf:53]:
```suggestion
variable "cloud_build_repositories" {
```

File: modules/cloudbuild_repo_connection/variables.tf [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+  validation {
+    condition     = alltrue([for k, v in var.cloudbuild_repos : try(length(regex("^https://.*\\.git$", v.repo_url)) > 0, false)])
+    error_message = "Each repo_url must be a valid HTTPS git clone URL ending with '.git'."
+  }
+
+}
+
+variable "default_region" {
+  description = "Default resources location"
```
Comment (daniel-cit) [inline modules/cloudbuild_repo_connection/variables.tf:68]:
```suggestion
variable "region" {
  description = "Resources location"
```

File: modules/cloudbuild_repo_connection/variables.tf [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    )
+    error_message = "You must specify a valid credential_type ('GITHUBv2' or 'GITLABv2'). For 'GITHUBv2', all 'github_' prefixed variables must be defined and no 'gitlab_' prefixed variables should be defined. For 'GITLABv2', all 'gitlab_' prefixed variables must be defined and no 'github_' prefixed variables should be defined."
+  }
+}
+
+variable "cloudbuild_repos" {
+  description = "Object structure to bring your own repositories."
+  type = map(object({
+    repo_name = string,
+    repo_url  = string,
```
Comment (daniel-cit) [inline modules/cloudbuild_repo_connection/variables.tf:57]:
```suggestion
    repository_name = string,
    repository_url  = string,
```

File: test/integration/cloudbuild_repo_connection_github/cloudbuild_repo_connection_github_test.go [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+		// validate if repository was created using the connection
+		project_id := bpt.GetTFSetupStringOutput("project_id")
+		location := "us-central1"
+		connection_id := bpt.GetStringOutput("cloudbuild_2nd_gen_connection")
+		connection_slice := strings.Split(connection_id, "/")
+
+		assert.True(len(connection_slice) > 0, "Connection ID should be in format projects/{{project}}/locations/{{location}}/connections/{{name}}")
+
+		connection_name := connection_slice[len(connection_slice)-1]
+		repository := gcloud.Run(t, fmt.Sprintf("builds repositories describe %s", repoName), gcloud.WithCommonArgs([]string{"--project", project_id, "--region", location, "--connection", connection_name, "--format", "json"}))
```
Comment (daniel-cit) [inline test/integration/cloudbuild_repo_connection_github/cloudbuild_repo_connection_github_test.go:126]: using `gcloud.RunF` would make the code easier to read

File: test/integration/cloudbuild_repo_connection_github/cloudbuild_repo_connection_github_test.go [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+		// validate if repository was created using the connection
+		project_id := bpt.GetTFSetupStringOutput("project_id")
+		location := "us-central1"
+		connection_id := bpt.GetStringOutput("cloud_build_repositories_2nd_gen_connection")
+		connection_slice := strings.Split(connection_id, "/")
+
+		assert.True(len(connection_slice) > 0, "Connection ID should be in format projects/{{project}}/locations/{{location}}/connections/{{name}}")
+
+		connection_name := connection_slice[len(connection_slice)-1]
+		repository := gcloud.Runf(t, "builds repositories describe %s --project %s, --region %s --connection %s", repoName, project_id, location, connection_name)
```
Comment (daniel-cit) [inline test/integration/cloudbuild_repo_connection_github/cloudbuild_repo_connection_github_test.go:126]:
```suggestion
		repository := gcloud.Runf(t, "builds repositories describe %s --project %s --region %s --connection %s", repoName, project_id, location, connection_name)
```

File: test/integration/cloudbuild_repo_connection_gitlab/cloudbuild_repo_connection_gitlab_test.go [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+		// validate if repository was created using the connection
+		project_id := bpt.GetTFSetupStringOutput("project_id")
+		location := "us-central1"
+		connection_id := bpt.GetStringOutput("cloud_build_repositories_2nd_gen_connection")
+		connection_slice := strings.Split(connection_id, "/")
+
+		assert.True(len(connection_slice) > 0, "Connection ID should be in format projects/{{project}}/locations/{{location}}/connections/{{name}}")
+
+		connection_name := connection_slice[len(connection_slice)-1]
+		repository := gcloud.Runf(t, "builds repositories describe %s --project %s, --region %s --connection %s", repoName, project_id, location, connection_name)
```
Comment (daniel-cit) [inline test/integration/cloudbuild_repo_connection_gitlab/cloudbuild_repo_connection_gitlab_test.go:140]:
```suggestion
		repository := gcloud.Runf(t, "builds repositories describe %s --project %s --region %s --connection %s", repoName, project_id, location, connection_name)
```

File: modules/cloudbuild_repo_connection/versions.tf [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+  required_providers {
+    google = {
+      source = "hashicorp/google"
+      # Exclude 4.31.0 for https://github.com/hashicorp/terraform-provider-google/issues/12226
+      version = ">= 4.17, != 4.31.0, < 6"
+    }
+
+    time = {
+      source  = "hashicorp/time"
+      version = "0.12.0"
```
Comment (apeabody) [inline modules/cloudbuild_repo_connection/versions.tf:29]:
```suggestion
      version = ">= 0.12.0"
```

File: modules/cloudbuild_repo_connection/versions.tf [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    random = {
+      source  = "hashicorp/random"
+      version = "3.6.2"
```
Comment (apeabody) [inline modules/cloudbuild_repo_connection/versions.tf:34]:
```suggestion
      version = ">= 3.6.2"
```

File: modules/cloudbuild_repo_connection/versions.tf [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+terraform {
+  required_version = ">=0.15.0"
```
Comment (apeabody) [inline modules/cloudbuild_repo_connection/versions.tf:18]: Needed for `optional()` in variables.tf

```suggestion
  required_version = ">=1.3"
```

File: examples/cloudbuild_repo_connection_gitlab/main.tf [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+module "gitlab_connection" {
+  source = "../../modules/cloudbuild_repo_connection"
```
Comment (apeabody) [inline examples/cloudbuild_repo_connection_gitlab/main.tf:18]: For examples please use the registry and version (e.g. https://github.com/terraform-google-modules/terraform-google-bootstrap/blob/master/examples/tf_cloudbuild_builder_simple/main.tf#L18), and the CI will automatically replace with this prior source line when running lint and tests.

File: test/integration/cloudbuild_repo_connection_gitlab/cloudbuild_repo_connection_gitlab_test.go [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+
+		connectionSlice := strings.Split(connectionId, "/")
+		// Extract the project, location and connection name from the connection_id
+		connectionProjectId := connectionSlice[1]
+		connectionLocation := connectionSlice[3]
+		connectionName := connectionSlice[len(connectionSlice)-1]
+
+		// Assert that the resource was created in the specified project and region
+		assert.Equal(projectId, connectionProjectId, "Connection project id should be the same as input project id.")
+		assert.Equal(resourcesLocation, connectionLocation, "Connection location should be 'us-central'.")
```
Comment (apeabody) [inline test/integration/cloudbuild_repo_connection_gitlab/cloudbuild_repo_connection_gitlab_test.go:149]: `us-central1`? Perhaps better to use `resourcesLocation` in `fmt.Sprintf`.

```suggestion
		assert.Equal(resourcesLocation, connectionLocation, "Connection location should be 'us-central'.")
```

File: test/integration/cloudbuild_repo_connection_github/cloudbuild_repo_connection_github_test.go [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+
+		connectionSlice := strings.Split(connectionId, "/")
+		// Extract the project, location and connection name from the connection_id
+		connectionProjectId := connectionSlice[1]
+		connectionLocation := connectionSlice[3]
+		connectionName := connectionSlice[len(connectionSlice)-1]
+
+		// Assert that the resource was created in the specified project and region
+		assert.Equal(projectId, connectionProjectId, "Connection project id should be the same as input project id.")
+		assert.Equal(resourcesLocation, connectionLocation, "Connection location should be 'us-central'.")
```
Comment (apeabody) [inline test/integration/cloudbuild_repo_connection_github/cloudbuild_repo_connection_github_test.go:135]: `us-central1`? Perhaps better to use `resourcesLocation` in `fmt.Sprintf`.

```suggestion
		assert.Equal(resourcesLocation, connectionLocation, "Connection location should be 'us-central`'.")
```

**Review by apeabody — CHANGES_REQUESTED** [review]
Thanks @caetano-colin!  A few minor things I noted during the final review.

**Comment by caetano-colin** [comment]
@apeabody thank you for the review! I implemented the requested changes. Regarding the module registry version I put:

```terraform
version = "~> 9.0"
```

I'm assuming that would be the module version, If that is not correct please let me know.

**Comment by apeabody** [comment]
> @apeabody thank you for the review! I implemented the requested changes. Regarding the module registry version I put:
> 
> ```hcl
> version = "~> 9.0"
> ```
> 
> I'm assuming that would be the module version, If that is not correct please let me know.

Yeah, that should be fine for now.  Thanks!

**Comment by apeabody** [comment]
@caetano-colin - The last INT test didn't complete before timeout, might want to bump slightly: https://github.com/terraform-google-modules/terraform-google-bootstrap/blob/master/build/int.cloudbuild.yaml#L15

**Review by apeabody — APPROVED** [review]

**Comment by apeabody** [comment]
@gtsorbo do you want to review before merging?

**Comment by caetano-colin** [comment]
> @caetano-colin - The last INT test didn't complete before timeout, might want to bump slightly: https://github.com/terraform-google-modules/terraform-google-bootstrap/blob/master/build/int.cloudbuild.yaml#L15

Created a commit that increases the timeout in 10minutes

**Review by daniel-cit — APPROVED** [review]
LGTM

**Pull request merged**